book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,497 @@
+
+- Location: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+impl UpdateWith<PreparedBlock> for ValueBalance<NegativeAllowed> {
+    fn update_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
+        self += value_balance?;
+    }
+
+    fn revert_chain_state_with(&mut self, value_balance: &ValueBalance<NegativeAllowed>) -> Result<(), Err> {
+        self -= value_balance?;
+    }
+}
+```
+
+### Changes to finalized state
+
+The state service will call `commit_new_chain()`. We need to pass the value pool from the disk into this function.
+
+```rust
+self.mem
+    .commit_new_chain(prepared, self.disk.history_tree(), self.disk.get_pool())?;
+```
+
+We now detail what is needed in order to have the `get_pool()` method available.
+
+#### Serialization of `ValueBalance<C>`
+
+In order to save `ValueBalance` into the disk database we must implement `IntoDisk` and `FromDisk` for `ValueBalance` and for `Amount`:
+
+```rust
+impl IntoDisk for ValueBalance<C> {
+    type Bytes = [u8; 32];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        [self.transparent.to_bytes(), self.sprout.to_bytes(),
+        self.sapling.to_bytes(), self.orchard.to_bytes()].concat()
+    }
+}
+
+impl FromDisk for ValueBalance<C> {
+    fn from_bytes(bytes: impl AsRef<[u8]>) -> Self {
+        let array = bytes.as_ref().try_into().unwrap();
+        ValueBalance {
+            transparent: i64::from_be_bytes(array[0..8]).try_into().unwrap()
+            sprout: i64::from_be_bytes(array[8..16]).try_into().unwrap()
+            sapling: i64::from_be_bytes(array[16..24]).try_into().unwrap()
+            orchard: i64::from_be_bytes(array[24..32]).try_into().unwrap()
+        }
+    }
+}
+
+impl IntoDisk for Amount {
+    type Bytes = [u8; 8];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.to_bytes()
+    }
+}
+
+impl FromDisk for Amount {
+    fn from_bytes(bytes: impl AsRef<[u8]>) -> Self {
+        let array = bytes.as_ref().try_into().unwrap();
+        i64::from_be_bytes(array).try_into().unwrap()
+    }
+}
+```
+
+#### Changes to `zebra-state/src/request.rs`
+
+Add a new field to `FinalizedState`:
+
+```rust
+pub struct FinalizedBlock {
+    ..
+    /// The value balance for transparent, sprout, sapling and orchard
+    /// inside all the transactions of this block.
+    pub(crate) block_value_balance: ValueBalance<NegativeAllowed>,
+}
+```
+
+Populate it when `PreparedBlock` is converted into `FinalizedBlock`:
+
+```rust
+impl From<PreparedBlock> for FinalizedBlock {
+    fn from(prepared: PreparedBlock) -> Self {
+        let PreparedBlock {
+            ..
+            block_value_balance,
+        } = prepared;
+        Self {
+            ..
+            block_value_balance,
+        }
+    }
+}
+```
+
+#### Changes to `zebra-state/src/service/finalized_state.rs`
+
+First we add a column for all the pools: transparent, sprout, sapling, orchard:
```

Comment:
Unclear names: Can we put "chain value pool" in every name in this section?

That will help us see the difference between chain value pools and block value balances.

Reply:
https://github.com/ZcashFoundation/zebra/pull/2430/commits/5ae19fc0d76e060313f03a99fe97a07740451522